src/Semantic/Git.hs

```diff
@@ -24,22 +30,40 @@ clone url path = sh $ do
 -- | git cat-file -p
 catFile :: FilePath -> OID -> IO Text
 catFile gitDir (OID oid) = sh $ do
-  run "git" [pack ("--git-dir=" <> gitDir), "cat-file", "-p", oid]
+  run "git" ["-C", pack gitDir, "cat-file", "-p", oid]
 
 -- | git ls-tree -rz
 lsTree :: FilePath -> OID -> IO [TreeEntry]
-lsTree gitDir (OID sha) = sh $ do
-  out <- run "git" [pack ("--git-dir=" <> gitDir), "ls-tree", "-rz", sha]
-  pure $ mkEntry <$> splitOn "\NUL" out
-  where
-    mkEntry row | [mode, ty, rest] <- splitOn " " row
-                , [oid, path] <- splitOn "\t" rest
-                = TreeEntry (objectMode mode) (objectType ty) (OID oid) (unpack path)
-                | otherwise = nullTreeEntry
+lsTree gitDir (OID sha) = sh $ parseEntries <$> run "git" ["-C", pack gitDir, "ls-tree", "-rz", sha]
 
 sh :: MonadIO m => Sh a -> m a
 sh = shelly . silently . onCommandHandles (initOutputHandles (`hSetBinaryMode` True))
 
+-- | Parses an list of entries separated by \NUL, and on failure return []
+parseEntries :: Text -> [TreeEntry]
+parseEntries text = case parseOnly everything text of
+  Done "" ls -> ls
+  other      -> error ("There was an error parsing the Git output: " <> show other)
+  where
+    everything = AP.sepBy entryParser "\NUL" <* ("\NUL\n" <?> "End sequence") <* AP.endOfInput <?> "Everything"
+    parseOnly p t = AP.feed (AP.parse p t) ""
+
+-- | Parse the entire input with entryParser, and on failure return a default
+-- For testing purposes only
+parseEntry :: Text -> TreeEntry
+parseEntry = either (const nullTreeEntry) id . AP.parseOnly (entryParser <* AP.endOfInput)
+
+-- | Parses a TreeEntry
+entryParser :: Parser TreeEntry
+entryParser = TreeEntry
+  <$> modeParser <* (" " <?> "First Space")
```

Comment:
```suggestion
  <$> modeParser <* char ' '
```

Error messages with attoparsec are mostly a lost cause, but there's no reason to annotate a space with an error message, given that the error message will say it expected a space anyway. I also think that using the `char` parser here is faster and a little more explicit than relying on `OverloadedStrings` here.

Reply:
When testing this code, I ended up debugging a parse output that failed on the parser "string", with no further context. Adding these annotations made me able to figure out which parser was failing. I'm still happy to remove them, but it will further degrade error messages.

Reply:
Makes sense. I don’t think we really need to worry about error messages for now, given the relatively normal shape of this data, now that you’ve debugged it, so I’d remove the annotations for the sake of readability. (`char ‘ ‘` will give a better error message than `string “ “`, by the way).